crates/autopilot/src/shadow.rs

```diff
@@ -119,30 +128,25 @@ impl RunLoop {
             .orders
             .set(i64::try_from(auction.orders.len()).unwrap_or(i64::MAX));
 
-        let mut participants = self.competition(id, auction).await;
-
-        // Shuffle so that sorting randomly splits ties.
-        participants.shuffle(&mut rand::thread_rng());
-        participants.sort_unstable_by_key(|participant| cmp::Reverse(participant.score()));
+        let participants = self.competition(id, auction).await;
+        let winners = self.select_winners(&participants);
 
-        if let Some(Participant {
-            driver,
-            solution: Ok(solution),
-        }) = participants.first()
-        {
-            let reference_score = participants
-                .get(1)
-                .map(|participant| participant.score())
+        for (i, Participant { driver, solution }) in winners.iter().enumerate() {
+            let reference_score = winners
+                .get(i + 1)
+                .map(|winner| winner.score())
+                .unwrap_or_default();
```

Comment:
It's unclear to me whether the last winner should have 0 or the next smaller `score` as the reference. 🤔

Reply:
Ah, good catch. Will need to update this.

Reply:
Actually, I'll have to consult @cowprotocol/solver crew about this.

Reply:
We have not finalized the design. How urgent is this?

Our current approach is as follows:
1. First choose all winners. This is done by iteratively choosing the batch with largest score from all remaining solutions and then filtering all auctions which have a (directed) token (pair) in common with that batch.
2. Second determine reference scores. For winner with index i this is done by looking at all solutions i was compared to (i.e. excluding solutions with overlap to a previous winner) except winner i+1 and choosing the largest one among those.
3. Lastly compute rewards. Rewards of the first winner  is the usual difference of winning and reference score capped at the usual cap. For winner i the differences of the score and reference score is capped at the payment of winner i-1.

Step 3 is not done by backend. I just added it to give an idea why we need to exclude winners in 2.: We want to cap payments such that the payment of winner i cannot be smaller than the payment of winner i+1. Then, if the payment of winner i+1 is capped by the payment of winner i, they could increase their payment by submitting a worse solution. Thus we exclude that solution for finding a reference.

@acanidio-econ

Reply:
> Second determine reference scores. For winner with index i this is done by looking at all solutions i was compared to (i.e. excluding solutions with overlap to a previous winner) except winner i+1 and choosing the largest one among those.

I am having trouble understanding this part even after reading it several times. Why doing `except winner i+1`?
So, let's say after scoring we have the ordering:
1. winner
2. non-winner (because it contains order with tokens that overlap with (1))
3. winner
4. non-winner (because it contains order with tokens that overlap with (1) or (3))

What is the reference score for (3)? Is it score of (4)?

Reply:
After discussion I decided to implement a default rule that for a winner (3), reference score is a score of non-winner (4). If we decide to change this before feature activation we will have a follow up PR to update the rules.

Reply:
> After discussion I decided to implement a default rule that for a winner

Did that discussion happen outside this thread or did you decide this because @fhenneke didn't respond anymore?